server/coordinator.go

```
@@ -0,0 +1,288 @@
+	defer timer.Stop()
+
+	for {
+		select {
+		case <-timer.C:
+			timer.Reset(ctrl.GetInterval())
+			if !ctrl.AllowSchedule() {
+				continue
+			}
+			if op := s.Schedule(c.cluster); op != nil {
+				c.addOperator(s.GetResourceKind(), op)
+			}
+		case <-ctrl.Done():
+			return
+		}
+	}
+}
+
+func (c *coordinator) addOperator(kind ResourceKind, op *balanceOperator) {
+	regionID := op.Region.GetId()
+	if _, ok := c.regionCache.get(regionID); ok {
+		return
+	}
+	if c.getOperator(regionID) != nil {
+		return
+	}
+	c.setOperator(kind, op)
+}
+
+func (c *coordinator) setOperator(kind ResourceKind, op *balanceOperator) {
+	c.Lock()
+	defer c.Unlock()
+
+	collectOperatorCounterMetrics(op)
+	c.operators[kind][op.Region.GetId()] = op
+}
+
+func (c *coordinator) removeOperator(op *balanceOperator) {
+	c.Lock()
+	defer c.Unlock()
+
+	regionID := op.Region.GetId()
+	c.histories.add(regionID, op)
+	c.regionCache.set(regionID, nil)
+
+	for _, ops := range c.operators {
+		delete(ops, regionID)
+	}
+}
+
+func (c *coordinator) getOperator(regionID uint64) *balanceOperator {
+	c.RLock()
+	defer c.RUnlock()
+
+	for _, ops := range c.operators {
+		if op, ok := ops[regionID]; ok {
+			return op
+		}
+	}
+
+	return nil
+}
+
+func (c *coordinator) getOperators() map[uint64]Operator {
+	c.RLock()
+	defer c.RUnlock()
+
+	operators := make(map[uint64]Operator)
+	for _, ops := range c.operators {
+		for id, op := range ops {
+			operators[id] = op
+		}
+	}
+
+	return operators
+}
+
+func (c *coordinator) getHistories() []Operator {
+	c.RLock()
+	defer c.RUnlock()
+
+	var operators []Operator
+	for _, elem := range c.histories.elems() {
+		operators = append(operators, elem.value.(Operator))
+	}
+
+	return operators
+}
+
+func (c *coordinator) getOperatorCount(kind ResourceKind) int {
+	c.RLock()
+	defer c.RUnlock()
+	return len(c.operators[kind])
+}
+
+// Controller is an interface to control the speed of different schedulers.
+type Controller interface {
```

Need to return the Err of context?